future-apps/payroll/test/contracts/Payroll.test.js

```diff
@@ -0,0 +1,3362 @@
+const { assertRevert } = require('@aragon/test-helpers/assertThrow')
+const { encodeCallScript } = require('@aragon/test-helpers/evmScript')
+const { getEvent, getEvents, getEventArgument } = require('../helpers/events')
+const { bn, bigExp, maxUint64, maxUint256, annualSalary } = require('../helpers/numbers')(web3)
+const { deployErc20TokenAndDeposit, redistributeEth, getDaoFinanceVault } = require('../helpers/setup.js')(artifacts, web3)
+
+const ACL = artifacts.require('ACL')
+const Payroll = artifacts.require('PayrollMock')
+const PriceFeed = artifacts.require('PriceFeedMock')
+const ExecutionTarget = artifacts.require('ExecutionTarget')
+
+const MAX_GAS_USED = 6.5e6
+const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'
+
+contract('Payroll, initialization,', ([owner, employee, anotherEmployee, anyone]) => {
+  let dao, acl, payroll, payrollBase, finance, vault, priceFeed, denominationToken, anotherToken
+
+  const NOW = 1553703809 // random fixed timestamp in seconds
+  const ONE_MONTH = 60 * 60 * 24 * 31
+  const TWO_MONTHS = ONE_MONTH * 2
+  const RATE_EXPIRATION_TIME = TWO_MONTHS
+
+  const PCT_ONE = bigExp(1, 18)
+  const DENOMINATION_TOKEN_DECIMALS = 18
+
+  const currentTimestamp = async () => payroll.getTimestampPublic()
+
+  before('setup base apps and tokens', async () => {
+    const contracts = await getDaoFinanceVault(owner)
+    dao = contracts.dao
+    finance = contracts.finance
+    vault = contracts.vault
+    acl = ACL.at(await dao.acl())
+
+    priceFeed = await PriceFeed.new()
+    payrollBase = await Payroll.new()
+
+    anotherToken = await deployErc20TokenAndDeposit(owner, finance, vault, 'Another token', 18)
+    denominationToken = await deployErc20TokenAndDeposit(owner, finance, vault, 'Denomination Token', DENOMINATION_TOKEN_DECIMALS)
+
+    await redistributeEth(finance)
+  })
+
+  beforeEach('create payroll instance', async () => {
+    const receipt = await dao.newAppInstance('0x4321', payrollBase.address, '0x', false, { from: owner })
+    payroll = Payroll.at(getEventArgument(receipt, 'NewAppProxy', 'proxy'))
+  })
+
+  beforeEach('set timestamps', async () => {
+    await priceFeed.mockSetTimestamp(NOW)
+    await payroll.mockSetTimestamp(NOW)
+  })
+
+  beforeEach('grant permissions', async () => {
+    const ADD_EMPLOYEE_ROLE = await payrollBase.ADD_EMPLOYEE_ROLE()
+    const ADD_ACCRUED_VALUE_ROLE = await payrollBase.ADD_ACCRUED_VALUE_ROLE()
+    const CHANGE_PRICE_FEED_ROLE = await payrollBase.CHANGE_PRICE_FEED_ROLE()
+    const MODIFY_RATE_EXPIRY_ROLE = await payrollBase.MODIFY_RATE_EXPIRY_ROLE()
+    const TERMINATE_EMPLOYEE_ROLE = await payrollBase.TERMINATE_EMPLOYEE_ROLE()
+    const SET_EMPLOYEE_SALARY_ROLE = await payrollBase.SET_EMPLOYEE_SALARY_ROLE()
+    const ALLOWED_TOKENS_MANAGER_ROLE = await payrollBase.ALLOWED_TOKENS_MANAGER_ROLE()
```

Comment:
Probably a super small thing, but since this gets called for every test and should be constant, could we cache their values in `before()`?